app/models.py

```diff
@@ -20,14 +21,155 @@
+    """
+    Create an instance of Session when you
+    want to apply a verification method on all added
+    or modified or deleted rows before commiting them
+    to the database.
+
+    This class overrides the commit() function
+    by adding a verify() function before it.
+
+    Use this in a context manager:
+    with VerifiedSession(user_object) as session:
+        ...
+        session.commit()
+
+    This will make sure the changes to the DB
+    are verified, and will close the connection
+    when the context goes out.
+
+    """
+
+    def __init__(self, user_or_token):
+        """
+        This session must be initialized with a user or token.
+        Get this token from the handler (`self.current_user`)
+        or be generating an unverified session to only query
+        the user with a certain id. Example:
+
+        with DBSession() as session:
+            user = session.scalars(
+                sa.select(User).where(User.id == user_id)
+            ).first()
+
+        Parameters
+        ----------
+        user_or_token: baselayer.app.models.User object
+            or baselayer.app.models.Token object.
+            The object representing the current user.
+
+        """
+        self.user_or_token = user_or_token
+
+    def verify(self):
+        """Check that the current user has permission to create, read,
+        update, or delete rows that are present in the session. If not,
+        raise an AccessError (causing the transaction to fail and the API to
+        respond with 401).
+
+        """
+        # get items to be inserted
+        new_rows = [row for row in self.new]
+
+        # get items to be updated
+        updated_rows = [row for row in self.dirty if self.is_modified(row)]
+
+        # get items to be deleted
+        deleted_rows = [row for row in self.deleted]
+
+        # get items that were read
+        read_rows = [
+            row
+            for row in set(self.identity_map.values())
+            - (set(updated_rows) | set(new_rows) | set(deleted_rows))
+        ]
+
+        # need to check delete permissions before flushing, as deleted records
+        # are not present in the transaction after flush (thus can't be used in
+        # joins). Read permissions can be checked here or below as they do not
+        # change on flush.
+        for mode, collection in zip(
+            ["read", "update", "delete"],
+            [read_rows, updated_rows, deleted_rows],
+        ):
+            bulk_verify(mode, collection, self.user_or_token)
+
+        # update transaction state in DB, but don't commit yet. this updates
+        # or adds rows in the database and uses their new state in joins,
+        # for permissions checking purposes.
+        self.flush()
+        bulk_verify("create", new_rows, self.user_or_token)
+
+    def commit(self):
+        print("verifying rows before committing... ")
+        self.verify()
+        super().commit()
+
+
+VerifiedSession = scoped_session(
```

Comment:
It's confusing to have Session and VerifiedSession being entirely different things. Maybe the class should be called `_VerifiedSession` to make clear that `VerifiedSession` is the one to use?

Reply:
Good call.